plugins/json_ld.ts

```diff
@@ -0,0 +1,102 @@
+import { merge } from "../core/utils/object.ts";
+import { getDataValue } from "../core/utils/data_values.ts";
+
+import type Site from "../core/site.ts";
+import type { Page } from "../core/file.ts";
+import { Graph, Thing } from "npm:schema-dts@1.1.2";
+export interface Options {
+  /** The list extensions this plugin applies to */
+  extensions?: string[];
+
+  /** The key name for the transformations definitions */
+  name?: string;
+}
+
+const defaults: Options = {
+  extensions: [".html"],
+  name: "json_ld",
+};
+export type JsonldData = Graph | Thing;
+
+// Helper function to determine if a value is an object
+function isObject(value: unknown) {
+  return value && typeof value === "object" && !Array.isArray(value);
+}
+
+/**
+ * A plugin to insert structured JSON-LD data for SEO and social media
+ * @see https://lume.land/plugins/json_ld/
+ */
+export function json_ld(userOptions?: Options) {
+  const options = merge(defaults, userOptions);
+
+  return (site: Site) => {
+    site.mergeKey(options.name, "object");
+    site.process(options.extensions, (pages) => pages.forEach(json_ld));
+
+    function json_ld(page: Page) {
+      let json_ld = page.data[options.name] as JsonldData | undefined;
+
+      if (!json_ld || !page.document) {
+        return;
+      }
+      const { document, data } = page;
+
+      // Recursive function to traverse and process JSON-LD data
+      function traverse(key: string | undefined, value: unknown): unknown {
```

Comment:
For clarity, this function should be moved out of `json_ld` (processor fn) because it's not neede to define it for every page.

Reply:
Sure, but the traverse function captures the data, so I’ll need to pass it to the function.

Reply:
Oh, you're right, I missed that.

In this case, I guess it's fine.